### PR TITLE
WT-10449 Do not save update chain when there are no updates to be written to the history store

### PR DESCRIPTION
**Pull request by kommiharibabu**

### CODE REVIEW

File: src/reconcile/rec_visibility.c [hunk 1]
```diff
@@ -910,10 +910,18 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, W
...
-        WT_RET(__rec_update_save(
-          session, r, ins, rip, onpage_upd, upd_select->tombstone, supd_restore, upd_memsize));
-
+        /*
+         * No need to save the update chain when it meets the following:
+         * 1. No need to restore the update chain
+         * 2. No on-disk entry exist
+         * 3. No further updates exist in the update chain to be written to the history store.
+         */
+        if (supd_restore || vpack != NULL || onpage_upd->next != NULL) {
```
Comment (quchenhao) [inline src/reconcile/rec_visibility.c:919]: Why not implement the logic in __rec_need_save_upd?

Comment (kommiharibabu) [reply src/reconcile/rec_visibility.c:919]: I tried that approach and it needs more changes than the current approach. As we need to set the update with the DS flag irrespective of whether we save the update chain except when it is globally visible. I can do that change if you find that way is good.

Comment (quchenhao) [reply src/reconcile/rec_visibility.c:919]: Can we move the code to set the ds flag outside this if section?

Comment (kommiharibabu) [reply src/reconcile/rec_visibility.c]: Yes, moved it outside and add the conditions inside the need_save_upd function.

File: src/reconcile/rec_visibility.c [hunk 2]
```diff
@@ -277,6 +279,18 @@ __rec_need_save_upd(
...
+    /*
+     * No need to save the update chain when it meets the following:
+     * 1. No need to restore the update chain
+     * 2. No on-disk entry exist
+     * 3. No further updates exist in the update chain to be written to the history store.
+     */
+    supd_restore =
+      F_ISSET(r, WT_REC_EVICT) && (has_newer_updates || F_ISSET(S2C(session), WT_CONN_IN_MEMORY));
+
+    if (!supd_restore && vpack == NULL && upd_select->upd != NULL && upd_select->upd->next == NULL)
```
Comment (quchenhao) [inline src/reconcile/rec_visibility.c:291]: Can we walk to next update that us not aborted?

Comment (kommiharibabu) [reply src/reconcile/rec_visibility.c]: Yes, added a loop to check for a valid update before saving the update chain.

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/reconcile/rec_visibility.c [hunk 3]
```diff
@@ -278,9 +281,33 @@ __rec_need_save_upd(
...
-        return (!__wt_txn_tw_stop_visible_all(session, &upd_select->tw));
+        visible_all = __wt_txn_tw_stop_visible_all(session, &upd_select->tw);
     else
-        return (!__wt_txn_tw_start_visible_all(session, &upd_select->tw));
+        visible_all = __wt_txn_tw_start_visible_all(session, &upd_select->tw);
+
+    if (visible_all)
+        return (false);
+
+    /*
```
Comment (sulabhM) [inline src/reconcile/rec_visibility.c:291]: Here is an alternate comment suggestion if that helps:

```
/*
 * Update chains are only saved if there are newer uncommitted updates or for the in-memory storage.
 * If there has not been an on-disk value before, a new key has been inserted. We need to save updates
 * if there is at least one update that has not been aborted.
 */
```

**Review by sulabhM — APPROVED** [review]
I have an alternate comment suggestion for you to explore. Otherwise, this change looks good.

File: src/reconcile/rec_visibility.c (hunk 3 above)
Comment (kommiharibabu) [reply src/reconcile/rec_visibility.c:291]: Thanks. Adjusted the comment to understand the need for saving the update chain.

**Pull request merged**